### DISCUSSION TITLE
Not collecting sub-modules

### DISCUSSION THREAD

**Discussion by Safihre** [pyinstaller/pyinstaller#8446] [Help] [answered]
I am trying to add `apprise` to our SABnzbd package. But for some reason the dynamic loading of AppRise is conflicting with PyInstaller logic.

AppRise specifies plugins in a `apprise.plugins` directory, which is then read during initialization.
This seams to happen here:
https://github.com/caronc/apprise/blob/master/apprise/manager.py#L62
https://github.com/caronc/apprise/blob/master/apprise/manager.py#L187
https://github.com/caronc/apprise/blob/master/apprise/manager.py#L205

Just to be sure, I added all sub-modules of AppRise to the `hiddenimports`, here:
https://github.com/sabnzbd/sabnzbd/blob/bugfix/release_issues/builder/SABnzbd.spec#L17
And it seems to pick them up (I can also see the binary size expand a lot):
https://github.com/sabnzbd/sabnzbd/actions/runs/8936839729/job/24547944375

But during execution, it fails to find the plugins:
```
2024-05-03 11:49:31,646::DEBUG::[manager:268] Notification Plugin 0(s) and 0 Schema(s) loaded in 0.0000s
```

I assume it has to do with paths. So what does AppRise or we have to do to make it work with PyInstaller?

CC: @caronc

**Comment by rokm** [comment, marked as answer]
It has to do the with the fact that plugins are discovered by listing the directory contents, while in PyInstaller-frozen build those plugin modules are in the PYZ archive.

You can work around it by ensuring that contents of `apprise.plugins` are collected as .py files as well.

Nowadays, the canonical way to do so is to set [`module_collection_mode`](https://pyinstaller.org/en/stable/hooks.html), which you can do either in a hook or in the spec file:
```python
a = Analysis(
    ...
    module_collection_mode={
        'apprise.plugins': 'pyz+py',  # or even just 'py'
    },
)
```

**Reply by Safihre** [reply]
Will this not break in MacOS, where Pyton might try to compile the py files, breaking the signature? 
Or setting `pyz+py` would prevent that?

**Reply by rokm** [reply]
The embedded interpreter is [set up with `write_bytecode=0`](https://github.com/pyinstaller/pyinstaller/blob/5663cd502ee5a346725e6a63f2d82405670c9996/bootloader/src/pyi_pyconfig.c#L579) so it should not be writing .pyc files.

But in this case, the .py files are used only to discover plugins (a `pkgutil.iter_modules` might be a better approach?), so with `pyz+py` I think the actual modules should end up loaded from PYZ.

**Reply by Safihre** [reply]
Thank you, your suggestion worked! 🥳
